target_chains/solana/programs/pyth-price-publisher/src/accounts/publisher_prices.rs

```diff
@@ -0,0 +1,173 @@
+    if data.len() < size_of::<u32>() {
+        return false;
+    }
+    let format: &u32 = from_bytes(&data[..size_of::<u32>()]);
+    *format == FORMAT
+}
+
+pub fn read(data: &[u8]) -> Result<(&PublisherPricesHeader, &[PublisherPrice]), ReadAccountError> {
+    if data.len() < size_of::<PublisherPricesHeader>() {
+        return Err(ReadAccountError::DataTooShort);
+    }
+    let header: &PublisherPricesHeader = from_bytes(&data[..size_of::<PublisherPricesHeader>()]);
+    if header.format != FORMAT {
+        return Err(ReadAccountError::FormatMismatch);
+    }
+    let prices_bytes = &data[size_of::<PublisherPricesHeader>()..];
+    let num_prices: usize = header.num_prices.try_into().unwrap();
+    let expected_len = num_prices.saturating_mul(size_of::<PublisherPrice>());
+    if expected_len > prices_bytes.len() {
+        return Err(ReadAccountError::InvalidNumPrices);
+    }
+    let prices = cast_slice(&prices_bytes[..expected_len]);
+    Ok((header, prices))
+}
+
+pub fn size(max_prices: usize) -> usize {
+    size_of::<PublisherPricesHeader>() + max_prices * size_of::<PublisherPrice>()
+}
+
+pub fn read_mut(
+    data: &mut [u8],
+) -> Result<(&mut PublisherPricesHeader, &mut [u8]), ReadAccountError> {
+    if data.len() < size_of::<PublisherPricesHeader>() {
+        return Err(ReadAccountError::DataTooShort);
+    }
+    let (header, prices) = data.split_at_mut(size_of::<PublisherPricesHeader>());
+    let header: &mut PublisherPricesHeader = from_bytes_mut(header);
+    if header.format != FORMAT {
+        return Err(ReadAccountError::FormatMismatch);
+    }
+    Ok((header, prices))
+}
+
+pub fn create(
+    data: &mut [u8],
+    publisher: [u8; 32],
+) -> Result<(&mut PublisherPricesHeader, &mut [u8]), ReadAccountError> {
+    if data.len() < size_of::<PublisherPricesHeader>() {
+        return Err(ReadAccountError::DataTooShort);
+    }
+    let (header, prices) = data.split_at_mut(size_of::<PublisherPricesHeader>());
+    let header: &mut PublisherPricesHeader = from_bytes_mut(header);
+    *header = PublisherPricesHeader::new(publisher);
+    Ok((header, prices))
+}
+
+pub fn extend(
+    header: &mut PublisherPricesHeader,
+    prices: &mut [u8],
+    new_prices: &[u8],
+) -> Result<(), ExtendError> {
+    if new_prices.len() % size_of::<PublisherPrice>() != 0 {
+        return Err(ExtendError::InvalidLength);
+    }
+    let num_new_prices = (new_prices.len() / size_of::<PublisherPrice>())
+        .try_into()
+        .expect("unexpected overflow");
+    let num_prices: usize = header.num_prices.try_into().unwrap();
+    let start = size_of::<PublisherPrice>() * num_prices;
+    let end = size_of::<PublisherPrice>() * num_prices + new_prices.len();
+    header.num_prices = header
+        .num_prices
+        .checked_add(num_new_prices)
+        .expect("unexpected overflow");
+    prices
+        .get_mut(start..end)
+        .ok_or(ExtendError::NotEnoughSpace)?
+        .copy_from_slice(new_prices);
```

Comment:
i get why we don't cast here and it is very good. to make this copy faster, let's use `sol_memcpt`. Also it's good to write this assumption that we don't do any input validation for the sake of speed here.

Reply:
Done.